### ISSUE TITLE
Pester crashes if user does not have alias `foreach` defined

### ISSUE THREAD

**Issue by ninmonkey** [pester/Pester#1740]
## 1. General summary of the issue

Invoke Pester crashes if the user's profile contains this:
```powershell
Remove-Alias -Name 'foreach' -Force
```

## 2. Describe Your Environment

```powershell
Pester version     : 5.0.4 C:\Users\cppmo_000\Documents\PowerShell\Modules\Pester\5.0.4\Pester.psd1
PowerShell version : 7.0.3
OS version         : Microsoft Windows NT 10.0.19041.0
```

## 3. Expected Behavior
`Invoke-Pester` should run without the user requiring user aliases.

## 4.Current Behavior

```powershell
Invoke-Pester

System.Management.Automation.CommandNotFoundException: The term 'foreach' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)    
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
at Invoke-Pester<End>, C:\Users\user\Documents\PowerShell\Modules\Pester\5.0.4\Pester.psm1: line 4501
at <ScriptBlock>, <No file>: line 
```

## 5. Possible Solution

Running `PSScriptAnalyzer` with `PSAvoidUsingCmdletAliases` to find the `ForEach-Object`s

**Comment by fflaten** [comment]
Shouldn't be using `Foreach-Object` direct either, so double issue here. I've seen a PR (#1640) focusing on some of this cleanup, but it looks to be inactive. Should probably solve some of these in a new PR for now.

**Comment by nohwnd** [comment]
There are also quite a few usages of Where-Object, Select, %, and sort.

**Comment by fflaten** [comment]
I have problems reproducing the error in 5.0.4 (think I got it once), but the line in the stacktrace is updated in pending PR #1745 . Need help to confirm when the PR is reviewed and hopefully approved. 🙂

**Comment by nohwnd** [comment]
@fflaten I am not sure I understand. Do you have hard time reproducing the error in the latest version? I saw it happen in 5.0.4 but not in the newer version, because there I already changed the code to use foreach keyword (not alias). I think your changes look good and I commented in few places.

**Comment by fflaten** [comment]
I tried to reproduce using PS 7.0.3 and Pester 5.0.4, but didn't get the exception. Probably user error on my part as I originally was able to reproduce one time. As long as you can confirm it works in latest version or using PR-code then I'm happy. 👍

**Comment by asears** [comment]
I went through my inactive PR #1640 to fix the PSScriptAnalyzer warns.  Still a few things outstanding there.

During testing and reviewing other issues, I found that some of the suggestions Script Analyzer provided such as Write-Host caused issues with the documentation tool and missing Write-Host parameters.  The behavior of Write-Host and Write-Output is different so it's not a 1-1 substitution.  So I scrapped those changes in the PR.

I also found odd failures when replacing foreach with the expanded Foreach-Object.  Would there be some difference in the alias vs. the expansion?  I think foreach may be used as alias for Foreach-Object, but also some other internal powershell commands.  Couldn't find anything definitive on that...

The draft PR is still open and I merged latest with v5.0 if someone wants to take code suggestions out of it or take over the PR.  I didn't feel comfortable that some of the changes would introduce unintended behaviour.

Is the use of &SafeCommands documented in the Pester documentation somewhere?  I see a single reference for adding it when adding new functions in pester.dev.  Might be worth adding additional docs around why this is used and how to best implement through the code, and perhaps a custom PSScriptAnalyzer rule.  My first instinct would have been to just replace the alias with the call to the command.

**Comment by fflaten** [comment]
@nohwnd
> I saw it happen in 5.0.4 but not in the newer version, because there I already changed the code to use foreach keyword (not alias).

Pester\5.0.4\Pester.psm1: line 4501 (from stacktrace) is:
`$containers += @(Find-File -Path $PesterPreference.Run.Path.Value -ExcludePath $PesterPreference.Run.ExcludePath.Value -Extension $PesterPreference.Run.TestExtension.Value | foreach { New-BlockContainerObject -File $_ })`

This wasn't modified in 5.1.0-beta* prior to #1745.
https://github.com/pester/Pester/blob/3e902a003e789a54dffebb157b41608700276f3e/src/Pester.ps1#L1026

Either way, it's updated now. Issue ready to close?

@asears Would prefer to discuss that in the PR or it's related issue #1635 so this issue can be closed without losing track of your questions. As for `$SafeCommands` I guess it's kinda documented now with the new analyzerrule introduced in #1745 
![image](https://user-images.githubusercontent.com/3436158/97786143-e2987f80-1ba9-11eb-8208-0a58f04e42ec.png)
